### ISSUE TITLE
subscriptions and stages need to seperated in to two kinds

### ISSUE THREAD

**Issue by tal-hason** [akuity/kargo#896]
# Summary

when creating a stage a subscription is part of it, but in fact, a subscription is a different identity that needs to be its own kind.

![image](https://github.com/akuity/kargo/assets/97454240/1eba12ee-b9a8-4f3e-9fca-f67e42fb0f31)

so the stage can register to a subscription, making the flow much more visible and manageable.

# Motivation

make the source of the repo/chart/image a subscription that stages can then be registered to, and then stages can also register to other upstream stages.

# Proposal

freight then should be a stage registered to a subscription, a stage without a subscription, will be able only to consume freights, not generate them, this will assist with more agility in designing the pipeline into different scenarios.

stage should look like:

```YAML
apiVersion: kargo.akuity.io/v1alpha1
kind: Stage
metadata:
  name: test
  namespace: kargo-demo
spec:
  upstreamStages:
    - name: <<name of the stage>>
  registerTo:
    subscriptions:
      - name: <<name of the subscription>>
  promotionMechanisms:
    argoCDAppUpdates:
    - appName: kargo-demo-test
      sourceUpdates:
      - repoURL: https://github.com/<username>/kargo-demo-gitops.git
        helm:
          images:
          - image: nginx
            key: image.name
            value: Image # Update Application to reference the new image version
```

Subscription should look like this:

```YAML
apiVersion: kargo.akuity.io/v1alpha1
kind:  Subscriptions
metadata:
  name:   subscriptions
  namespace: kargo-demo
spec:
  repos:  #atleast one type of repo is required for a subscription
      images: 
      - repoURL: nginx
        semverConstraint: ^1.24.0
      charts:
      - registryURL: https://charts.bitnami.com/bitnami
        name: nginx 
        semverConstraint: ^14.1.0
      git:
      - repoURL: https://github.com/<username>/kargo-demo-gitops.git
        branch: helm # Watch this branch for new commits
```

**Comment by krancour** [comment]
This is sort of an incidental duplicate of #631, which proposes a Freight CRD and, perhaps more importantly, decoupling Freight production from the "zero stages." (This is an unofficial shorthand we use to denote stages on the far left of a pipeline that subscribe directly to repositories.)

Somewhere in the thread, this is said:

> * We would need a way for users to define something that produces Freight (e.g. FreightFactory)
> * Stages would no longer have syntax to subscribe to image/git/helm explicitly but rather a selector of freight. Though, I still like the syntax of subscribing to upstream Stages.

Note the name "FreightFactory" will not necessarily stick, but it got the point across. Whatever we call it, I believe this is _exactly_ what you're requesting here, so I am going to close this as a duplicate of #631 -- only because that's the issue we're already working from. _Work on this is already underway._

I invite you to please re-open this issue if you disagree that the proposal here and #631 are less similar than I understand them to be.

**Comment by tal-hason** [comment]
@krancour , yes I understand and agree with most if not all what is written in #631 , but I just what to make a not that I think that is missing there.

From what I understand/feel freight should be a bundle of sources(images/helm/git) and a combination of those should create a freight.

So a stage that is subscribe to sources will start a freight chain and upstream stages will will be able to promote/convey to it.

So there will be 3 types of kinds??
1. Source/subscription,  a single image helm and git kind
2. FreightGenerator/freightliners combination of source/subscription.
3. Stages, consumer of freights or other stages.

Hope this make some sense

**Comment by krancour** [comment]
> So there will be 3 types of kinds??
>
>    1. Source/subscription, a single image helm, and git kind
>    2.  FreightGenerator/Freightliner combination of source/subscription.
>    3. Stages, consumers of freights or other stages.

No. 1 doesn't seem necessary at this time. Decoupling the production of Freight from the "zero stages" is already a humongous pivot. I'd like to get that done before introducing _more_ CRDs into the equation.

Can you explain why you think no. 1 and no. 2 need to be separate things?

**Comment by tal-hason** [comment]
@krancour ,from my view point.
We are talking about kargos (cargos) and freight and Freightliner.

So delivery hub for cargos with freights.

What goes in the freight can be made from different sources, and if I have a source that can be found in 2 freight, there no reason to query/ create this resource twice.

Look at my pictures it really explain the idea.

**Comment by krancour** [comment]
@tal-hason I don't want to be dismissive of ideas, but I want to propose to you that this isn't really necessary.

If you could imagine the repo(s) subscription section that is currently part of a Stage spec being lifted out of there and put in it's own CRD, that is what #631 is going for. You've left with something that is decoupled from the "zero stage" and it can assemble/create Freight from multiple sources.

It is the idea that each _source_ needs to be a resource that I am not convinced of the necessity of.

Compare the following bits of equivalent pseudo-YAML:

```
Source:
  Name: my-image-repo
  URL: my-image-url
---
Source:
  Name: your-image-repo
  URL: your-image-url
---
FreightFactory:
  Name: our-freight-factory
  Sources:
  - my-image-repo
  - your-image-repo
```

vs

```
FreightFactory:
  Name: our-freight-factory
  Sources:
  - my-image-url
  - your-image-url
```

Either would produce the same Freight, but one is clearly more complicated than the other. I am struggling to understand what value that extra bit of complexity is supposed to be adding here.

fwiw, nothing would prevent it from being visualized exactly as you depicted.

**Comment by krancour** [comment]
I'll re-open this until we understand if there really is something here that's different from #631.

**Comment by tal-hason** [comment]
@krancour , if we are just moving the zero-stage to a freight-factory, there isn't much change except that now stages will be registered to freight id's instead of a stage that holds that freight-id.

This will still requires duplicating repos if there is a need to build several deployments stages from a combination of sources.

Think of sources as your storage dispenser, you have couple of those, now you want to create different delivery packages for different stages, i.e.

Developer a, is working on part 1 but needs to get all other parts of the system.
Developer b is working on part 2 but needs to get all other parts of the system.

Development env. Needs to accumulate both part 1 and 2. And from there continue the uat and prod.

I'm my idea you only have 2 sources and 3 frieght factorys

Vs. 

The freight-factory only you will have 3 freight-factory's with 4 source's

So the controller will have more reconciliation loops/sampling for those repos. Twice as much.

And with more images or repos it can be multiplied many more.

**Comment by krancour** [comment]
> there isn't much change except that now stages will be registered to freight id's instead of a stage that holds that freight-id.

This isn't true. It's a _huge_ change.

Before #631 is done, Freight isn't a CRD. It's something that exists _only_ as a status entry in each Stage. Eventually, every stage will forget old pieces of Freight and then they're effectively gone forever. Worse, the ability to query, filter, sort Freight, etc. requires jumping through lots and lots of hoops. It's terribly inefficient at present.

#631 will formalize each piece of Freight as a resource -- a more tangible thing that can _easily_ be queried, filtered, sorted, and even garbage collected according to user-defined policies -- so that old Freight that is, for some reason, especially important, can be retained for longer periods.

If Freight becomes a CRD, _something_ like the "FreightFactory" has to _produce_ it.

If each piece of Freight then references the "Factory" that created it, every piece of Freight then has an implicit "type." The notion of a Freight type could prevent unsupported scenarios. The other day, you tried to merge multiple parallel pipelines and that's something we don't support (_yet_). If your final Stage had expressed that it expected Freight of a certain type, which was _different_ from the types of Freight your upstream Stages had been expecting, the invalidity of what you were attempting would have been immediately clear.

So Freight and "FreightFactory" CRDs buy us a lot.

The thing I am _still_ not convinced of is that we need resources to represent individual repositories.

1. I don't want the rules for _what_ makes its way in the door to be defined at the "Repository" level. Imagine that you have two different pipelines that need the same image, but one wants to move every new major release of the image through the pipeline and the other wants to move every new patch release through. What I really _don't_ want to do is have to define two _separate_ "Repo" resources where one uses one set of subscription rules and the other uses another. _I want the "Factory" that assembles Freight from upstream sources to decide on these rules._

1. Taking the above into consideration, if the subscription "rules" are defined at the "Factory" level, then a hypothetical "Repo" resource is almost nothing more than a simple reference to some remote repo. If that's _all_ it is and not anything more, easier for "Factories" to reference repos by URL than by a reference to another resource. i.e. I don't want a new resource type where a _string_ would do the trick.

1. Last, I am averse to additional complexities until they prove themselves absolutely necessary. The idea of making Kargo too complicated to understand is a genuine concern.

**Comment by tal-hason** [comment]
@krancour , thanks, I get your point, and I get, that the freight will hold the sources and their needed version/tag to create the unique freight ID, and I can agree with that and also live with that.

Just want to highlight #879 which regards using patterns and placeholders for image details.

so how this will be incorporated with a FreightFactory? if at all 

thanks again for the detailed answer :smiley:

**Comment by krancour** [comment]
I think this doesn't factor into #879 

#879 proposes patterns and placeholders within certain promotion mechanisms. Promotion mechanisms are mostly a mapping from an artifact to an action. "When you see this image, here's what I want you to do with it." #879 doesn't change that. "When you see an image that matches this pattern, here's what I want you to do with it." I think that still works fine after these changes.

What I will concede to is #631 (or #631 + the extra CRD you proposed; with or without doesn't matter here) combined with the pattern/matching support creates a strange possibility that a "FreightFactory" could begin subscribing to an additional artifact that it didn't subscribe to before and this might be matched by a regex in a stage and now the Stage is doing a new promotion step that it wasn't before and perhaps didn't intend. That could be avoided if we _didn't_ do #631, but #631 is a priority for too many other reasons.

Priority-wise #631 comes first. I am already working on it. We can re-evaluate after how that affects the pattern/placeholder proposal and how to mitigate it. It's also possible that we decide the issue is too esoteric to deserve any special consideration. Since "FreightFactories" would be scoped to the same project/namespace as the Stages that consume them, I am perhaps overstating the likelihood that someone changes a Factory without anticipating its downstream repercussions.

At any rate, we're starting to go off on a tangent that we can worry about later when it's time to deal with #879.

For now, I am satisfied that the original issue in this thread is indeed a duplicate of #631 and am going to close this, but we'll be open-minded about re-opening this if we discover after #631 that something still seems to be missing.
